### PR TITLE
Documentation for OAuth2.0 friendly login

### PR DESCRIPTION
**Pull request by kishansagathiya**

### CODE REVIEW

File: docs/source/reference.adoc [hunk 1]
```diff
@@ -124,3 +124,68 @@ __Response:__
 
 [source,json]
 {"access_token":"eyJhbGciOiJSUzI1NiIsImtpZCI6IjlNTG5WaWFSa2hWajFHVDlrcFdVa3dISXdVRC13WmZVeFItM0Nwa0UtWHMiLCJ0eXAiOiJKV1QifQ.eyJpYXQiOjE1MTA0ODg4NTMsImlzcyI6Imh0dHA6Ly9sb2NhbGhvc3Q6ODA4OSIsImp0aSI6IjAzNmE0OTM2LTEwNTEtNDQxMC05MTI1LTg4YWM1ODE5NDc1NiIsInNjb3BlcyI6WyJ1bWFfcHJvdGVjdGlvbiJdLCJzZXJ2aWNlX2FjY291bnRuYW1lIjoiZmFicmljOC13aXQiLCJzdWIiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUifQ.box9XsE5XR1-OaRdAsKneDabeaOAtqZXcAbFDrwSqYNvwm_zoNEUmG1lDvq7DVVTdRXSvm8dASx0nGOBwrC_Iv2DXlOsO2DxO1lSMDfeQwtFLE9jbwFlqW7xmcjtQLkWqgzGr5gpV-bZo7-lxQuHv4A_hZAiWv03K6hU7qD57KfWuGQm71gV5rqP_Rm3Ay2HDzgmEBiaYoh6XGS5jw4tk-8X6wAJF9phJ73qAQ0E8OcbM9JTPQUQnrbTuaHW-CmtQ4_1TYaBP_j1K__C-o0K14BEGJvyj3geU8CvGjFl7EL94YZmGHidOZtCgRGjNR0KbD6yXUlPMALEhT2R2j65qw","token_type":"bearer"}
+
+== OAuth2.0 Friendly Login
+
+=== Steps for User Login:
+
+. Client sends a generated state. Auth "proxies" it to KC and thereby initiates the first in oauth2
```
Comment (sbose78) [inline docs/source/reference.adoc]: 'initiates the first.. ?

File: docs/source/reference.adoc [hunk 2]
```diff
@@ -124,3 +124,68 @@ __Response:__
...
+GET /api/authorize
+        response_type=code
+        client_id=740650a2-9c44-4db5-b067-a3d1b2cd2d01
+        scope=user:email
+        state=18184535-097d-473e-8dce-0688952d8439
+        redirect_uri=https://auth.prod-preview.openshift.io/api/status
+
+required fields: response_type, client_id, state
+
+- _Response:_
```
Comment (sbose78) [inline docs/source/reference.adoc]: add an example.

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-264-1`

File: docs/source/reference.adoc [hunk 3]
```diff
@@ -124,3 +124,68 @@ __Response:__
...
+[source]
+POST /api/token
+        grant_type=authorization_code
+        client_id=740650a2-9c44-4db5-b067-a3d1b2cd2d01
+        code=uss.gZs_vR-oBxlzUsYxesh71gBWIZfjYSxSP6xjC8T-WCM.12ca5270-c30f-4c41-90f8-ccbfc5fca498.98447a27-40de-450b-a987-1ddc97730839
+        redirect_uri=https://auth.prod-preview.openshift.io/api/status
+
+required fields: all
+
+- _Response:_
```
Comment (sbose78) [inline docs/source/reference.adoc]: add an example.

File: docs/source/reference.adoc [hunk 4]
```diff
@@ -124,3 +124,68 @@ __Response:__
 
 [source,json]
 {"access_token":"eyJhbGciOiJSUzI1NiIsImtpZCI6IjlNTG5WaWFSa2hWajFHVDlrcFdVa3dISXdVRC13WmZVeFItM0Nwa0UtWHMiLCJ0eXAiOiJKV1QifQ.eyJpYXQiOjE1MTA0ODg4NTMsImlzcyI6Imh0dHA6Ly9sb2NhbGhvc3Q6ODA4OSIsImp0aSI6IjAzNmE0OTM2LTEwNTEtNDQxMC05MTI1LTg4YWM1ODE5NDc1NiIsInNjb3BlcyI6WyJ1bWFfcHJvdGVjdGlvbiJdLCJzZXJ2aWNlX2FjY291bnRuYW1lIjoiZmFicmljOC13aXQiLCJzdWIiOiI1ZGVjNWZkYi0wOWUzLTQ0NTMtYjczZi01YzgyODgzMmIyOGUifQ.box9XsE5XR1-OaRdAsKneDabeaOAtqZXcAbFDrwSqYNvwm_zoNEUmG1lDvq7DVVTdRXSvm8dASx0nGOBwrC_Iv2DXlOsO2DxO1lSMDfeQwtFLE9jbwFlqW7xmcjtQLkWqgzGr5gpV-bZo7-lxQuHv4A_hZAiWv03K6hU7qD57KfWuGQm71gV5rqP_Rm3Ay2HDzgmEBiaYoh6XGS5jw4tk-8X6wAJF9phJ73qAQ0E8OcbM9JTPQUQnrbTuaHW-CmtQ4_1TYaBP_j1K__C-o0K14BEGJvyj3geU8CvGjFl7EL94YZmGHidOZtCgRGjNR0KbD6yXUlPMALEhT2R2j65qw","token_type":"bearer"}
+
+== OAuth2.0 Friendly Login
+
+=== Steps for User Login:
```
Comment (sbose78) [inline docs/source/reference.adoc]: You should rename this section as "How it works" and move it to the end. 
For a client who's plugging into our oauth2.0 friendly login, "KC" is just an implementation detail.

File: docs/source/reference.adoc [hunk 5]
```diff
@@ -124,3 +124,68 @@ __Response:__
...
+
+. Client sends a generated state. Auth "proxies" it to KC and thereby initiates the first in oauth2
+. KC redirects to IdP etc. User logs in..
+. KC calls back Auth with a code
+. Auth passes the code to Client ( UI or another services )
+. Client then calls KC to do the code \<\-> token exchange.
+. Auth works as a smart proxy here between client and KC.
+
+=== To get authorization_code
```
Comment (sbose78) [inline docs/source/reference.adoc]: This two step process could be in the section "Steps to login"

**Comment by fabric8cd** [comment]
@kishansagathiya snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-264-2`

File: docs/reference.html [hunk 6]
```diff
@@ -633,10 +647,189 @@ <h3 id="_service_account_authentication">Service Account Authentication</h3>
...
+<td class="tableblock halign-left valign-top"><p class="tableblock">scope</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">scope of permission</p></td>
+</tr>
+<tr>
+<td class="tableblock halign-left valign-top"><p class="tableblock">state</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">random unique string generate by the person who calls this api to be safe from Cross Site Request Forging</p></td>
+</tr>
+<tr>
+<td class="tableblock halign-left valign-top"><p class="tableblock">redirect_uri</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">uri where you want to be redirect along with the token</p></td>
```
Comment (sbose78) [inline docs/reference.html]: URI should be in capital.

File: docs/reference.html [hunk 7]
```diff
@@ -633,10 +647,189 @@ <h3 id="_service_account_authentication">Service Account Authentication</h3>
...
+<td class="tableblock halign-left valign-top"><p class="tableblock">client_id</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">The client ID</p></td>
+</tr>
+<tr>
+<td class="tableblock halign-left valign-top"><p class="tableblock">scope</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">scope of permission</p></td>
+</tr>
+<tr>
+<td class="tableblock halign-left valign-top"><p class="tableblock">state</p></td>
+<td class="tableblock halign-left valign-top"><p class="tableblock">random unique string generate by the person who calls this api to be safe from Cross Site Request Forging</p></td>
```
Comment (sbose78) [inline docs/reference.html]: In the description column everything should start with a capital letter.

File: docs/source/reference.adoc [hunk 8]
```diff
@@ -124,3 +124,77 @@ __Response:__
...
+
+== OAuth2.0 Friendly Login
+
+=== How it works:
+
+. Client sends a generated state. Auth "proxies" it to Keycloak and thereby initiates the first request in oauth2
+. Keycloak redirects to IdP etc. User logs in..
+. Keycloak calls back Auth with a code
+. Auth passes the code to Client ( UI or another services )
+. Client then calls KC to do the code \<\-> token exchange.
```
Comment (sbose78) [inline docs/source/reference.adoc]: Avoid using the term "KC", call it keycloak.  ( in other places too )

File: docs/reference.html (hunk 7 above)
Comment (sbose78) [reply docs/reference.html]: "Random unique string to be generated by the client to protect from CSRF"

File: docs/reference.html (hunk 6 above)
Comment (sbose78) [reply docs/reference.html]: "URI where you to be redirected.."

**Pull request merged**